### ISSUE TITLE
Consistent Entity History API

### ISSUE THREAD

**Issue by jarohen** [xtdb/xtdb#830]
Currently, we have four functions on the Crux API to get hold of history:

- `(crux/history node eid)`
  Eagerly returns all of the history of the given entity, with corrections, in index order.
- `(crux/history-range node eid vt-start tt-start vt-end tt-end)`
  Eagerly returns all of the history of the entity within the time co-ordinates, with corrections.
- `(crux/[open-,]history-seq-[a,de]scending db eid)`
  These four are on the DB, and also return the documents, but no corrections.

There's a couple of issues with this (beyond having a number of functions doing a similar but slightly different thing):
- history-seq-ascending on the DB starts from the valid-time/tx-time in the DB.
  This is the opposite behaviour of the other functions on the DB (for which the DB's vt/tt are upper bounds) - if I want the history since the start of time in ascending order, say, I have to get a DB at the start of time then call history-seq-ascending (and I can't then specify an upper bound without a `take-while`, or equivalent).
  More importantly, though, it also means that subsequent calls return more data, as the node indexes more transactions - the DB is not a value, in this case.
- (see #827) The `history-range` function on the node are vulnerable to the same issue as #546 - if I call it with a tt-end the node hasn't yet indexed, and then call it with the same parameters once the node's indexed the transaction, I'll get different results.

Broadly speaking, we're looking for a history API where users can:
- request data eagerly or lazily
- request corrections or no corrections
- request ascending or descending
- request docs or no docs

Ideally, this would also:
- follow the principles of 'db as a value' - i.e. if I call the same function with the same params on a db with the same params, I'm guaranteed to get the same result.
- return the same structure regardless of the flags the user supplies - although additive would also be fine (i.e. the difference between the user requesting docs and not requesting docs might be the addition of a `:crux/doc` key).

One such return value might be:
```clojure
[{:crux.db/valid-time ...
  :crux.tx/tx-time ...
  :crux.tx/tx-id ...
  :crux.db/content-hash ...
  :crux/doc {...}}
 ...]
```

Corrections, if requested, are included inline - the sequence is sorted first by valid-time, then tx-time.

Regarding _where_ this API lives - given we'd like to preserve both the DB's properties (repeatable queries, and ensuring that the node has indexed all of the transactions), we'd prefer to put this function on the DB. We'll accept `:start` (bitemp co-ordinate, inclusive) and `:end` (bitemp co-ordinate, exclusive) maps on the entity history call. This does have the tradeoff that the DB already contains one vt/tt - for the history API, we'll use this vt/tt as an upper bound to whatever the user provides (to the `:start`, in a descending history, and the `:end` in an ascending history). Separately, we'd also appreciate it if `(crux/db node ...)` would take a map of vt + tt rather than positional params (because it's valid to pass tt but not vt) - one for another day.

With this in mind - we'd like to keep the `entity-history` signature as close as possible to how DB _currently_ works (and then, if we change DB later, we can change this too). Proposal:

```clojure
(crux/entity-history db eid :asc
                     {:with-docs? true
                      :with-corrections? true
                      :start {:crux.db/valid-time ...
                              :crux.tx/tx-time ...}
                      :end {:crux.db/valid-time ...
                            :crux.tx/tx-time ...}})
```

We'd also add `open-entity-history`, with lazy semantics.

A nuanced difference here is that the behaviour of the order-ascending request changes to fix the above issue - this would return the history from the `:start` vt/tt provided up to the `:end` vt/tt, or the vt/tt of the DB, whichever's earlier.

Thoughts?

**Comment by refset** [comment]
In regards corrections, this is a good opportunity to review whether/how to coalesce consecutive repeated vt versions when `corrections? false` (related: https://github.com/juxt/crux/issues/326), and equally how/whether to represent (collapse?) them when `corrections? true`.

**Comment by jarohen** [comment]
thanks @refset - on reflection, agreed. have implemented it that way in #841
